### PR TITLE
[node-forge]add forge.pki.createCertificate() and supplement interface Certificate

### PR DESCRIPTION
**Pull request by x22x22**
[node-forge]add forge.pki.createCertificate() and supplement interface Certificate

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/x22x22/forge/blob/74b7472b2143735fca0c68a7f20ba32c4fbd5397/lib/x509.js#L910
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/node-forge/index.d.ts [hunk 1]
```diff
@@ -107,11 +107,66 @@ declare module "node-forge" {
...
             publicKey: any;
             md: any;
+            setSubject:
+            /**
+              * Sets the subject of this certificate.
+              *
+              * @param attrs the array of subject attributes to use.
+              * @param uniqueId an optional a unique ID to use.
+              */
+            (attrs: any, uniqueId?: any) => {};
```
Comment (westy92) [inline types/node-forge/index.d.ts]: Define functions like the rest of the file.  Put a return type if possible, and stricter parameter types as well.

Comment (westy92) [reply types/node-forge/index.d.ts]: Add some "tests" for this new functionality.

**Review by westy92 — CHANGES_REQUESTED** [review]

File: types/node-forge/index.d.ts (hunk 1 above)
Comment (x22x22) [reply types/node-forge/index.d.ts]: I am a typecript novice, can you give an example of these two requirements?

Comment (westy92) [reply types/node-forge/index.d.ts]: Line 103 above is a good example.
```ts
addField(attr: CertificateField): void;
```
Something similar for this method, although untested and using an educated guess, may be:
```ts
setSubject(attrs: string[], uniqueId?: number): void;
```

As for "testing," just provide some example usage in the `node-forge-tests.ts` in this directory and make sure it compiles.

Comment (x22x22) [reply types/node-forge/index.d.ts]: ok, thank you!

Comment (x22x22) [reply types/node-forge/index.d.ts]: done, pls check.

File: types/node-forge/index.d.ts [hunk 2]
```diff
@@ -107,11 +107,60 @@ declare module "node-forge" {
             extensions: any[];
             publicKey: any;
             md: any;
+            /**
+             * Sets the subject of this certificate.
+             *
+             * @param attrs the array of subject attributes to use.
+             * @param uniqueId an optional a unique ID to use.
+             */
+            setSubject(attrs: any[], uniqueId?: string): void;
```
Comment (westy92) [inline types/node-forge/index.d.ts]: `attrs: CertificateField[]`?

File: types/node-forge/index.d.ts [hunk 3]
```diff
@@ -107,11 +107,60 @@ declare module "node-forge" {
...
+            /**
+              * Sets the subject of this certificate.
+              *
+              * @param attrs the array of subject attributes to use.
+              * @param uniqueId an optional a unique ID to use.
+              */
+            setIssuer(attrs: any[], uniqueId?: string): void;
```
Comment (westy92) [inline types/node-forge/index.d.ts]: `attrs: CertificateField[]`?

File: types/node-forge/index.d.ts [hunk 4]
```diff
@@ -107,11 +107,60 @@ declare module "node-forge" {
...
+            /**
+              * Sets the extensions of this certificate.
+              *
+              * @param exts the array of extensions to use.
+              */
+            setExtensions(exts: any[]): void
```
Comment (westy92) [inline types/node-forge/index.d.ts]: Define and use a `CertificateExtension` type?
I see a bunch of fields here: https://github.com/digitalbazaar/forge/blob/74b7472b2143735fca0c68a7f20ba32c4fbd5397/lib/x509.js#L2086

It looks like you might have to use tagged union types (https://blog.mariusschulz.com/2016/11/03/typescript-2-0-tagged-union-types)

```ts
type CertificateExtension = CertificateExtensionKeyUsage /* | ... */;
interface CertificateExtensionKeyUsage {
  name: 'keyUsage';
  digitalSignature?: string; // type?
  // ...
}
```

File: types/node-forge/index.d.ts [hunk 5]
```diff
@@ -107,11 +107,60 @@ declare module "node-forge" {
...
+            /**
+             * Gets an extension by its name or id.
+             *
+             * @param options the name to use or an object with:
+             *          name the name to use.
+             *          id the id to use.
+             *
+             * @return the extension or null if not found.
+             */
+            getExtension(options: string | {name: string}):{} | undefined
```
Comment (westy92) [inline types/node-forge/index.d.ts]: Something more like:
```ts
getExtension(options: string | {name: string;} | {id: number;}): CertificateExtension;
```

File: types/node-forge/index.d.ts [hunk 6]
```diff
@@ -107,11 +107,60 @@ declare module "node-forge" {
...
+            /**
+             * Signs this certificate using the given private key.
+             *
+             * @param key the private key to sign with.
+             * @param md the message digest object to use (defaults to forge.md.sha1).
+             */
+            sign(key: pki.Key, md: md.MessageDigest): void
```
Comment (westy92) [inline types/node-forge/index.d.ts]: Make sure to use semicolons at the end of lines.

**Review by westy92 — CHANGES_REQUESTED** [review]

File: types/node-forge/index.d.ts [hunk 7]
```diff
@@ -107,11 +107,60 @@ declare module "node-forge" {
...
+            /**
+             * Attempts verify the signature on the passed certificate using this
+             * certificate's public key.
+             *
+             * @param child the certificate to verify.
+             *
+             * @return true if verified, false if not.
+             */
+            verify(child: Certificate):boolean
```
Comment (westy92) [inline types/node-forge/index.d.ts]: `verify(child: Certificate): boolean;`

File: types/node-forge/index.d.ts (hunk 4 above)
Comment (x22x22) [reply types/node-forge/index.d.ts]: digitalSignature is number.
'CertificateExtension' is too complicated, I do not have time to add the complete.

**Review by westy92 — APPROVED** [review]

**Pull request merged**